crates/deltalake-core/src/delta_datafusion/cdf/scan.rs

```diff
@@ -0,0 +1,318 @@
+                    });
+                }
+
+                if !remove_actions.is_empty() {
+                    remove_files.push(CdcDataSpec {
+                        version,
+                        timestamp: ts.unwrap_or(0),
+                        actions: remove_actions,
+                        commit_info: commit_info.cloned(),
+                    });
+                }
+            }
+        }
+
+        Ok((change_files, add_files, remove_files))
+    }
+
+    fn get_add_action_type() -> Option<ScalarValue> {
+        Some(ScalarValue::Utf8(Some(String::from("insert"))))
+    }
+
+    pub async fn scan(&self) -> DeltaResult<SendableRecordBatchStream> {
+        let (cdc, add, _remove) = self.determine_files_to_read().await?;
+        let ctx = SessionContext::new();
+        register_store(self.log_store.clone(), ctx.state().runtime_env().clone());
+
+        let schema_fields: Vec<Field> = self
+            .schema
+            .all_fields()
+            .into_iter()
+            .filter(|f| !self.partition_values.contains(f.name()))
+            .cloned()
+            .collect();
+
+        let this_partition_values = self
+            .partition_values
+            .iter()
+            .map(|name| self.schema.field_with_name(name).map(|f| f.to_owned()))
+            .collect::<Result<Vec<_>, ArrowError>>()?;
+
+        // Setup for the Read Schemas of each kind of file, CDC files include commit action type so they need a slightly
+        // different schema than standard add file reads
+        let cdc_file_schema = create_cdc_schema(schema_fields.clone(), true);
+        let add_file_schema = create_cdc_schema(schema_fields, false);
+
+        // Setup the mapping of partition columns to be projected into the final output batch
+        // cdc for example has timestamp, version, and any table partitions mapped here.
+        // add on the other hand has action type, timestamp, version and any additional table partitions because adds do
+        // not include their actions
+        let mut cdc_partition_cols = CDC_PARTITION_SCHEMA.clone();
+        let mut add_partition_cols = ADD_PARTITION_SCHEMA.clone();
+        cdc_partition_cols.extend_from_slice(&this_partition_values);
+        add_partition_cols.extend_from_slice(&this_partition_values);
+
+        // Setup the partition to physical file mapping, this is a mostly unmodified version of what is done in load
+        let cdc_file_groups =
+            create_partition_values(self.schema.clone(), cdc, &self.partition_values, None)?;
+        let add_file_groups = create_partition_values(
+            self.schema.clone(),
+            add,
+            &self.partition_values,
+            Self::get_add_action_type(),
+        )?;
+
+        // Create the parquet scans for each associated type of file. I am not sure when we would use removes yet, but
+        // they would be here if / when they are necessary
+        let cdc_scan = ParquetFormat::new()
```

Comment:
Have you considered ctx.read_parquet?

Reply:
You lose some control when you use that facility and I'm not even sure that this would work given it looks like it's more expecting a directory and not explicit file paths. It might work, I haven't explored it extensively to be honest, but at a glance it doesn't look like it would provide the control necessary.

Reply:
It does accept a list of filepaths. Comparing the two, it looks like these params are the same (I think)

ParquetReadOptions.schema  == FileScanConfig.file_schema
table_paths == FileScanConfig.file_groups
ParquetReadOptions.table_partition_cols == FileScanConfig.table_partition_cols 

Project and limit I guess you do after read_parquet on the dataframe (I assume it's lazy).

The only gap is the statistics part, but you don't seem to use it.  Could be perhaps interesting to check if there is any performance difference between the two?